# Sesión 09 · Resume y conecta tablas

Calcula indicadores por grupos, incorpora catálogos y trabaja con fechas.

Duración: 150 minutos de clase + 45–60 minutos de tarea. Ejecuta en orden. Los ejemplos son independientes salvo indicación.

## Objetivos
- Agrupar y calcular indicadores.
- Unir tablas sin multiplicar filas.
- Resumir datos por período.

## Agrupar para responder una pregunta

groupby divide registros según una o varias claves. Una agregación resume cada grupo con sum, mean, count u otra operación. Elige el resumen según la pregunta.

Sumar cantidades devuelve unidades; contar filas devuelve registros. Son indicadores diferentes, aunque a veces tengan el mismo valor.

In [1]:
import pandas as pd

ventas = pd.DataFrame({"Producto": ["Mouse", "Mouse", "Teclado"],
                      "Cantidad": [3, 1, 2], "Ingreso": [47.97, 15.99, 51.0]})
resumen = ventas.groupby("Producto", as_index=False).agg(
    unidades=("Cantidad", "sum"),
    ingreso=("Ingreso", "sum"),
    registros=("Producto", "size")
)
print(resumen)

  Producto  unidades  ingreso  registros
0    Mouse         4    63.96          2
1  Teclado         2    51.00          1


**Qué observar:** Mouse: 4 unidades, 63.96 de ingreso y 2 registros. Teclado: 2, 51.00 y 1.

**Recuerda:** Un promedio de precios por registro no es lo mismo que el precio promedio ponderado por unidades.

## Unir tablas con una clave

merge incorpora información de otra tabla usando claves compartidas. Una unión left conserva todas las filas de la tabla izquierda y añade coincidencias de la derecha.

Si un catálogo tiene claves repetidas, una unión puede multiplicar filas e inflar totales. validate="many_to_one" exige que la clave de la derecha sea única. indicator permite revisar coincidencias.

In [2]:
import pandas as pd

ventas = pd.DataFrame({"Producto": ["Mouse", "Mouse", "Teclado"], "Cantidad": [3, 1, 2]})
catalogo = pd.DataFrame({"Producto": ["Mouse", "Teclado"], "Categoria": ["Accesorios", "Accesorios"]})
unidas = ventas.merge(catalogo, on="Producto", how="left", validate="many_to_one", indicator=True)
print(unidas)
print(len(ventas), len(unidas))

  Producto  Cantidad   Categoria _merge
0    Mouse         3  Accesorios   both
1    Mouse         1  Accesorios   both
2  Teclado         2  Accesorios   both
3 3


**Qué observar:** Tres filas antes y después; todas las coincidencias son both.

**Recuerda:** Comprueba claves, filas y totales antes y después de una unión. Una tabla que se ve bien puede contener filas multiplicadas.

## Fechas y tablas de resumen

to_datetime convierte texto en fechas. Especificar el formato reduce ambigüedades: 03/04 puede significar días distintos según la convención. errors="coerce" permite detectar fechas inválidas.

dt.to_period("M") obtiene el mes. pivot_table reorganiza datos y agrega valores. fill_value=0 es apropiado aquí solo porque el ejemplo define ausencia de registro como ausencia de ventas registradas.

In [3]:
import pandas as pd

ventas = pd.DataFrame({"Fecha": ["2026-01-05", "2026-01-20", "2026-02-02"],
                      "Producto": ["Mouse", "Teclado", "Mouse"], "Ingreso": [48, 51, 32]})
ventas["Fecha"] = pd.to_datetime(ventas["Fecha"], format="%Y-%m-%d", errors="coerce")
ventas["Mes"] = ventas["Fecha"].dt.to_period("M").astype(str)
tabla = ventas.pivot_table(index="Mes", columns="Producto", values="Ingreso", aggfunc="sum", fill_value=0)
print(tabla)

Producto  Mouse  Teclado
Mes                     
2026-01      48       51
2026-02      32        0


**Qué observar:** Enero: Mouse 48, Teclado 51. Febrero: Mouse 32, Teclado 0.

**Recuerda:** Un período incompleto no debe compararse sin aclaración con otro completo.

## Práctica
Intenta resolver antes de consultar las soluciones del aula web.

### Total por producto

Agrupa A, B, A con ingresos 100, 200, 50. Ordena los totales de mayor a menor.

**Criterio de logro:** B: 200; A: 150.

In [ ]:
# Escribe tu solución aquí.

### Detecta falta de correspondencia

Une ventas de A y B con un catálogo que solo contiene A. Identifica el producto sin categoría.

**Criterio de logro:** B tiene categoría faltante.

In [ ]:
# Escribe tu solución aquí.

## Tarea · Del registro al indicador

Trabaja con ventas.csv.

1. Calcula Ingreso y agrupa por Producto para sumar unidades e ingresos.
2. Crea un catálogo con categorías para Mouse, Teclado y Monitor; únelo con validación many_to_one.
3. Convierte Fecha con formato explícito y resume ingreso por día.
4. Comprueba que el total siga siendo 416.95 antes y después de la unión.

Evaluación: corrección 50 %, explicación 30 %, orden y reproducibilidad 20 %.

In [ ]:
# Desarrolla tu tarea en esta celda y agrega las que necesites.

## Mi conclusión
Escribe tu interpretación y una limitación.